src/main/java/InfrastructureManager/Master.java

```diff
@@ -1,37 +1,138 @@
-        String in;
-        String mapping;
-        while (true) {
-            in = master.fromInput();
-            mapping = master.execute(in);
-            master.toOutput(mapping);
-            if (in.equals("exit")) {break;}
+    /**
+     * Method for starting the main runner thread, declared in the config file
+     */
+    //TODO:This could maybe go in configurator
+    public void startMainRunner() {
+        for (Runner runner : runnerList) {
+            if (runner.getName().equals("Main")) {
+                running.add(runner);
+                new Thread(runner,"MainRunner").start();
+                break;
+            }
+        }
+    }
+
+    /**
+     * Method that given a certain Scenario, runs the corresponding ScenarioRunner and adds
+     * it to the list of running Runners. If called on a running ScenarioRunner, it will restart
+     * it
+     * @param scenario Scenario to be run
+     */
+    public void runScenario(Scenario scenario) {
+        ScenarioRunner scenarioRunner = getRunner(scenario);
+        if (running.contains(scenarioRunner)) { //If running then stop it and re-run
+            scenarioRunner.exit();
+            running.remove(scenarioRunner);
+        }
+        scenarioRunner.setScenario(scenario);
+        running.add(scenarioRunner);
+        new Thread(scenarioRunner).start(); // Run the scenario in another thread
+
+    }
+
+    /**
+     * Method to pause a running ScenarioRunner, given the scenario that is running
+     * @param scenario Running scenario to be paused
+     */
+    public void pauseScenario(Scenario scenario) {
+        ScenarioRunner scenarioRunner = getRunner(scenario);
+        if (running.contains(scenarioRunner)) {
+            scenarioRunner.pause();
         }
     }
 
+    /**
+     * Method to resume a paused ScenarioRunner, given its scenario
+     * @param scenario Paused scenario to be resumed
+     */
+    public void resumeScenario(Scenario scenario) {
+        ScenarioRunner scenarioRunner = getRunner(scenario);
+        if (running.contains(scenarioRunner)) {
+            scenarioRunner.resume();
+        }
+    }
+
+    /**
+     * Method for finding a ScenarioRunner in the Runner list, based on the scenario is
+     * supposed to run according to the configuration file
+     * @param scenario Scenario for which to find the ScenarioRunner
+     * @return ScenarioRunner that will run the given scenario
+     * @throws IllegalArgumentException If there is no scenarioRunner configured to run the
+     * given scenario
+     */
+    private ScenarioRunner getRunner(Scenario scenario) {
+        ScenarioRunner scenarioRunner;
+        for (Runner runner : runnerList) {
+            try {
+                scenarioRunner = (ScenarioRunner) runner;
+                if (scenarioRunner.getScenarioName().equalsIgnoreCase(scenario.getName())){
+                    return scenarioRunner;
+                }
+            } catch (ClassCastException e) {
```

Comment:
As a general rule, most of the time a general catch is better than specific ones.
Here we are missing the case where scenario is null, just change to catch (Exception e).